drivers/arm-cca/ukrsi/rsi.c

```diff
@@ -0,0 +1,288 @@
+	value[4] = args.a5;
+	value[5] = args.a6;
+	value[6] = args.a7;
+	value[7] = args.a8;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_realm_config(unsigned long config)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_REALM_CONFIG;
+	args.a1 = config;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_version(void)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_VERSION;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+void uk_rsi_init(struct ukplat_bootinfo *bi)
+{
+	struct rsi_realm_config __attribute__((aligned(PAGE_SIZE))) config;
+	const struct ukplat_memregion_desc *mrd;
+	__paddr_t pstart, pend, len;
+	unsigned long version;
+	unsigned long ret;
+	__u32 i;
+
+	version = uk_rsi_version();
+	uk_pr_info("RSI version %lu.%lu\n", (version >> 16),
+		   (version & 0xFFFF));
+
+	for (i = 0; i < bi->mrds.count; i++) {
+		mrd = &bi->mrds.mrds[i];
+
+		pstart = ALIGN_UP(mrd->pbase, __PAGE_SIZE);
+		len = ALIGN_UP(mrd->len, __PAGE_SIZE);
+		pend   = pstart + len;
+
+		ret = uk_rsi_setup_memory(pstart, pend, RSI_RAM);
+		if (ret != RSI_SUCCESS) {
+			uk_pr_err("RSI setup memory failed\n");
+			return;
+		}
+	}
+
+	ret = uk_rsi_realm_config((unsigned long)&config);
+	if (ret != RSI_SUCCESS) {
+		uk_pr_err("RSI realm config failed\n");
+		return;
+	}
+
+	uk_pr_info("RSI realm config ipa_width %lu, hash_algo %lu:\n",
+		   config.ipa_width, config.hash_algo);
+
+	/* set the mask of the unprotected bit */
+	unprotected_mask = (1UL) << (config.ipa_width - 1);
+
+	uk_pr_info("RSI initialized unprotected_mask %lx\n", unprotected_mask);
+}
+
+unsigned long
+uk_rsi_setup_memory(unsigned long base, unsigned long end, unsigned char ripas)
+{
+	unsigned long ret;
+	unsigned long new_base;
+
+	/* Iterate over the memory space to set RIPAS */
+	while (base != end) {
+		ret = uk_rsi_ipa_state_set(base, end, ripas, 0, &new_base);
+		base = new_base;
+		if (ret != RSI_SUCCESS)
+			break;
+	}
+	return ret;
+}
+
+void uk_rsi_setup_device(void)
+{
+	int rc;
+	unsigned long pages, prot;
+
+	/* set up device memory region */
+	pages = DIV_ROUND_UP(RAM_BASE_ADDR, PAGE_SIZE);
+	prot = PAGE_ATTR_PROT_RW | PAGE_NS_SHARED;
+
+	rc = ukplat_page_set_attr(ukplat_pt_get_active(), 0, pages, prot, 0);
+	if (unlikely(rc))
+		UK_CRASH("Failed to set up device memory region\n");
+}
+
+unsigned long uk_generate_attestation_token(unsigned long addr,
+					    unsigned long challenge[8],
+					    unsigned long *size)
+{
+	unsigned long ret;
+
+	ret = uk_rsi_attestation_token_init(addr, challenge);
+
+	if (ret != RSI_SUCCESS)
+		return ret;
+
+	do {
+		ret = uk_rsi_attestation_token_continue(addr, size);
+	} while (ret == RSI_INCOMPLETE);
+
+	return ret;
+}
+
+int set_memory_protected(unsigned long addr, int numpages)
+{
+	unsigned long ret;
+	unsigned long prot;
+	unsigned long phys_addr;
+
+	prot = PAGE_ATTR_PROT_RW;
+	ret = ukplat_page_set_attr(ukplat_pt_get_active(), addr, numpages, prot,
+				   0);
+
+	phys_addr = ukplat_virt_to_phys(addr);
```

Comment:
You need a cast here otherwise you'll get a GCC warning:
```
-       phys_addr = ukplat_virt_to_phys(addr);
+       phys_addr = ukplat_virt_to_phys((void *)addr);
```